File: solutions/0007-reverse-integer/solution.java
class Solution {
    public int reverse(int x) {
        int res = 0;

        while(x != 0) {
            int n = x%10;

            // positive overflow
            if(res > Integer.MAX_VALUE/10 || (res == Integer.MAX_VALUE/10 && n > 7)) return 0;

            // negative overflow
            if(res < Integer.MIN_VALUE/10 || (res == Integer.MIN_VALUE/10 && n < -8)) return 0;

            res = (res*10) + n ;
            x /= 10;
        }

        return res;        
    }
}

File: solutions/0007-reverse-integer/solution.cpp
class Solution {
public:
    int reverse(int x) {

        bool isNeg = x < 0;
        
        long long n = abs((long long)x);
        long long rev = 0;
        
        while(n > 0) {
            int r = n % 10;
            rev = rev * 10 + r;
            n /= 10;
            
            if(rev > INT_MAX) return 0;
        }
        
        return isNeg ? -rev : rev;
    }
};
